File: 2_medium/00034_find_first_and_last_position_of_element_in_sorted_array/solution.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int l = std::lower_bound(nums.begin(), nums.end(), target) - nums.begin();

        if (l == nums.size() || nums[l] != target) {
            return vector<int>{-1, -1};
        }

        int r = std::upper_bound(nums.begin(), nums.end(), target) - nums.begin() - 1;

        return vector<int>{l, r};
    }
};

File: 2_medium/00034_find_first_and_last_position_of_element_in_sorted_array/solution.rs
use std::cmp::Ordering;

impl Solution {
    pub fn search_range(nums: Vec<i32>, target: i32) -> Vec<i32> {

        let l = nums.binary_search_by(|num| match num.cmp(&target) {
            Ordering::Less => Ordering::Less,
            _ => Ordering::Greater,
        }).unwrap_or_else(|x| x);

        if l == nums.len() || nums[l] != target {
            return vec![-1, -1];
        }

        let r = nums.binary_search_by(|num| match num.cmp(&target) {
            Ordering::Less | Ordering::Equal => Ordering::Less,
            Ordering::Greater => Ordering::Greater,
        }).unwrap_or_else(|x| x) - 1;

        vec![l as i32, r as i32]
    }
}
